include/pmacc/particles/algorithm/CallForEach.hpp

```diff
@@ -0,0 +1,84 @@
+/* Copyright 2019 Rene Widera
+ *
+ * This file is part of PMacc.
+ *
+ * PMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with PMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "pmacc/Environment.hpp"
+#include "pmacc/particles/algorithm/ForEach.hpp"
+#include "pmacc/particles/frame_types.hpp"
+#include "pmacc/functor/Interface.hpp"
+
+
+namespace pmacc
+{
+namespace particles
+{
+namespace algorithm
+{
+
+    /** Meta functor to execute an operation on all particles
+     *
+     * @tparam T_SpeciesTypeOperator an operator to create the used species
+     *                           with the species type as ::type
+     * @tparam T_FunctorOperator an operator to create a particle functor
+     *                           with the functor type as ::type
+     */
+    template<
+        typename T_SpeciesTypeOperator,
+        typename T_FunctorOperator
+    >
+    struct CallForEach
+    {
+        /** Operate on the domain CORE and BORDER
+         *
+         * @param currentStep current simulation time step
+         */
+        HINLINE void
+        operator()( uint32_t const currentStep )
+        {
+            using SpeciesType = typename T_SpeciesTypeOperator::type;
+            using FrameType = typename SpeciesType::FrameType;
+
+            // be sure the species functor follows the pmacc functor interface
+            using UnaryFunctor = pmacc::functor::Interface<
+                typename T_FunctorOperator::type,
+                1u,
+                void
+            >;
+
+            DataConnector &dc = Environment<>::get().DataConnector();
+            auto speciesPtr = dc.get< SpeciesType >(
```

Comment:
(Subjective and not important): I would prefer to have `auto & species = *dc< ...` here.

Reply:
but references/alias are dangerous and always dangling.

Reply:
Well, same as raw pointers?

Reply:
Here we dereference ist some lines later, so there is no difference between both methods but I will change it back.